recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -0,0 +1,162 @@
+import argparse
+import os
+import sys
+import hashlib
+import stat
+import pwd
+import grp
+import fs_permissions_shared
+
+# Function to return tree instead of variable since it contains functions inside its structure
+def known_permissions_tree():
+    return {
+        '.': system_dir,
+        'boot': {
+            '.': system_dir,
+            'bootmode': system_file,
+            'grub': {
+                '.': system_dir,
+                '*': system_file,
+                'grubenv': permissions(0o664, 'admin', 'ni', FT_REG),
+                'grubenv.bak': permissions(0o664, 'admin', 'ni', FT_REG),
+                'recoverytool-ni-version': permissions(0o0444, 'admin', 'administrators', FT_REG)
+            }, 'runmode': {
+                '.': system_dir,
+                '*': system_file,
+                'bzImage': system_link(f'bzImage-{bash("$(uname -r)")[0]}'),
+            }
+        }, 'lib': {
+            '.': system_dir,
+            'modules': {
+                '.': system_dir,
+                '$(uname -r)': {
+                    '.': system_dir,
+                    '**': system_hier,
```

Comment:
Unfortunately, some files under here are executable. The function here will need to accommodate that.

You can use `find /lib/modules/$(uname -r) -type f -perm /111` to discover what those files are.

Perhaps a couple higher-order functions like so could help:

```python
def subset_permissions(included_expectation, excluded_expectation, is_included):
    def ret(path, stats, logger, md5sum):
        nonlocal included_expectation, excluded_expectation, is_included
        if is_included(path):
            return included_expectation(path, stats, logger, md5sum)
        else:
            return excluded_expectation(path, stats, logger, md5sum)
    return ret

def tree_contains(prefix, tree):
    def ret(path):
        # TODO possibly remove prefix, do tree lookup here
    return ret
```

Then, this line could be something like this:

```suggestion
                    '**': subset_permissions(
                        system_executable,
                        system_heir,
                        tree_contains(
                            f'/lib/modules/{bash("$(uname -r)")}',
                            {
                                'build': {
                                    'scripts: {
                                        'checkversion.pl': () # or whatever marker
                                    }
                                }
                            })),
```

(Maybe it'd be simpler to use a list than a tree, but since we're already traversing a tree elsewhere that might be more congruent.)

Reply:
Rather than doing a tree lookup, there's also the option of using heuristics (e.g. file extensions, presence of shebangs or ELF headers) to determine whether a file ought to be executable in a simpler way.

Reply:
Do we want to actually keep track of exactly which files are executable?

If not, we can simply add to system_hier to allow files with the executable bit.

If so, I think we could just do something like:

```
{
  ...
  'lib': {
    ...
    'modules': {
      ...
      '$(uname -r)': {
        '.': system_dir,
        '**': system_hier,
        'build': {
          'scripts': {
            '+(*.sh|*.pl)': permissions(0o0755, 'admin', 'administrators', FT_REG),
            'prune-kernel': permissions(0o0755, 'admin', 'administrators', FT_REG)
...
```

Since, if we don't list something, it automatically gets caught by `**`, and vice-versa; if we add a specific permission it will automatically override what gets added with the `**`. The precedence is set by order in the tree.

In other words:

1. `**` gets fleshed out into all the files and they get added to a dictionary mapping their path to `system_hier`.
2. `+(*.sh|*.pl)` gets fleshed out and any matching files replace their entries in the dict with the 0755 permissions check.
3. The specific file listed, `prune-kernel` will also replace its entry in the dict.

I think because of the override and extglob matching, we don't necessarily need to add further logic, although I could wrap `permissions(0o0755, 'admin', 'administrators', FT_REG)` into a `system_script` or `system_executable` function a la `system_dir` and `system_file`

Reply:
Sounds like that'll work; give it a try.

Reply:
I just noticed that if you do `bash -O extglob -c "echo /lib/modules/$(uname -r)/**"` it evaluates to the same as just `*`.

Turns out we also need to add the `globstar` extension. Which is why I missed that there were executables not being processed